### PR TITLE
Fixes to MATFeaturizer and Tests

### PR DESCRIPTION
**Pull request by atreyamaj**
# Pull Request Template

## Description
Fixes to MATFeaturizer and Tests. The output will now be compatible with the MATModel layers, which was not the case earlier.

## Type of change

Please check the option that is related to your PR.

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
  - In this case, we recommend to discuss your modification on GitHub issues before creating the PR
- [x] Documentations (modification for documents)

## Checklist

- [x] My code follows [the style guidelines of this project](https://deepchem.readthedocs.io/en/latest/development_guide/coding.html)
  - [x] Run `yapf -i <modified file>` and check no errors (**yapf version must be  0.22.0**)
  - [x] Run `mypy -p deepchem` and check no errors
  - [x] Run `flake8 <modified file> --count` and check no errors
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New unit tests pass locally with my changes
- [x] I have checked my code and corrected any misspellings

### CODE REVIEW

**Comment by rbharath** [comment]
It looks like some of the changes broke the CI. Could you take a look to see what might be causing the errors? Will do a thorough review once CI errors are fixed :)

**Comment by atreyamaj** [comment]
> It looks like some of the changes broke the CI. Could you take a look to see what might be causing the errors? Will do a thorough review once CI errors are fixed :)

The CI breakages are due to `import rdkit` in the try-except block.

**Comment by rbharath** [comment]
@atreyamaj It looks like we have some test failures from this PR:

```
self = <deepchem.feat.tests.test_mat_featurizer.TestMATFeaturizer testMethod=test_mat_featurizer>

    def test_mat_featurizer(self):
      """
      Test featurizer.py
      """
      featurizer = MATFeaturizer()
      out = featurizer.featurize(self.mol)
      assert (type(out) == np.ndarray)
      assert (out.shape == (1,))
>     assert (out[0].node_features.shape == (1, 3, 36))
E     AssertionError: assert (3, 36) == (1, 3, 36)
E       At index 0 diff: 3 != 1
E       Right contains one more item: 36
E       Full diff:
E       - (1, 3, 36)
E       ?  ---
E       + (3, 36)

deepchem/feat/tests/test_mat_featurizer.py:27: AssertionError
```

Could you take a look and fix these errors?

File: deepchem/feat/molecule_featurizers/mat_featurizer.py [hunk 1]
```diff
@@ -28,18 +51,37 @@ class MATFeaturizer(MolecularFeaturizer):
   This class requires RDKit to be installed.
   """
 
-  def __init__(
-      self,
-      one_hot_formal_charge: bool = True,
```
Comment (rbharath) [inline deepchem/feat/molecule_featurizers/mat_featurizer.py:33]: Why are we removing this flag?

File: deepchem/feat/molecule_featurizers/mat_featurizer.py [hunk 2]
```diff
@@ -2,6 +2,29 @@
 from deepchem.utils.molecule_feature_utils import one_hot_encode
 from deepchem.utils.typing import RDKitMol, RDKitAtom
 import numpy as np
+from typing import Tuple, Any
+from dataclasses import dataclass
+
+
+@dataclass
```
Comment (rbharath) [inline deepchem/feat/molecule_featurizers/mat_featurizer.py:9]: Could you tell me a bit more about the dataclass annotation? We haven't used this before in DeepChem.

File: deepchem/feat/molecule_featurizers/mat_featurizer.py [hunk 3]
```diff
@@ -28,18 +51,37 @@ class MATFeaturizer(MolecularFeaturizer):
...
+
+    Returns
+    ----------
+    mol: RDKitMol
+      A processed RDKitMol objeect which is embedded, UFF Optimized and has Hydrogen atoms removed. If the former conditions are not met and there is a value error, then 2D Coordinates are computed instead.
     """
+    try:
+      from rdkit.Chem import AllChem
+      from rdkit import Chem
+    except:
```
Comment (rbharath) [inline deepchem/feat/molecule_featurizers/mat_featurizer.py:74]: This is too broad a try/except. We should lilkely only catch a `ModuleNotFound` error

File: deepchem/feat/molecule_featurizers/mat_featurizer.py [hunk 4]
```diff
@@ -55,24 +97,115 @@ def atom_features(self, atom: RDKitAtom) -> np.ndarray:
...
+      adj_matrix = m
+
+    if dist_matrix is not None:
+      m = np.full((dist_matrix.shape[0] + 1, dist_matrix.shape[1] + 1), 1e6)
+      m[1:, 1:] = dist_matrix
+      dist_matrix = m
+
+    return node_features, adj_matrix, dist_matrix
+
+  def pad_array(self, array: np.ndarray, shape: Any) -> np.ndarray:
```
Comment (rbharath) [inline deepchem/feat/molecule_featurizers/mat_featurizer.py:170]: This should likely be a private method I think

File: deepchem/feat/molecule_featurizers/mat_featurizer.py [hunk 5]
```diff
@@ -55,24 +97,115 @@ def atom_features(self, atom: RDKitAtom) -> np.ndarray:
...
+      RDKit Mol object.
+
+    Returns
+    ----------
+    Atom_features: ndarray
+      Numpy array containing atom features.
+    """
+    return np.array([self.atom_features(atom) for atom in mol.GetAtoms()])
+
+  def add_dummy_node(
```
Comment (rbharath) [inline deepchem/feat/molecule_featurizers/mat_featurizer.py:131]: Should this be a private method? Not sure if any DeepChem users would want to directly call this method

File: deepchem/feat/molecule_featurizers/mat_featurizer.py [hunk 6]
```diff
@@ -55,24 +97,115 @@ def atom_features(self, atom: RDKitAtom) -> np.ndarray:
...
+    ----------
+    array: np.ndarray
+      Array padded to input shape.
+    """
+    result = np.zeros(shape=shape)
+    slices = tuple(slice(s) for s in array.shape)
+    result[slices] = array
+    return result
+
+  def pad_sequence(self, sequence: np.ndarray) -> np.ndarray:
```
Comment (rbharath) [inline deepchem/feat/molecule_featurizers/mat_featurizer.py:191]: This should probably be a private method

**Review by rbharath — COMMENTED** [review]
Did a first review pass here. We need to make tests pass in addition to these comments :)

File: deepchem/feat/molecule_featurizers/mat_featurizer.py (hunk 3 above)
Comment (atreyamaj) [reply deepchem/feat/molecule_featurizers/mat_featurizer.py]: Changed :+1:

File: deepchem/feat/molecule_featurizers/mat_featurizer.py [hunk 7]
```diff
@@ -28,18 +50,37 @@ class MATFeaturizer(MolecularFeaturizer):
...
     ----------
-    one_hot_formal_charge: bool, default True
-      If True, formal charges on atoms are one-hot encoded.
+    mol: RDKitMol
+      RDKit Mol object.
+
+    Returns
+    ----------
+    mol: RDKitMol
+      A processed RDKitMol objeect which is embedded, UFF Optimized and has Hydrogen atoms removed. If the former conditions are not met and there is a value error, then 2D Coordinates are computed instead.
```
Comment (rbharath) [inline deepchem/feat/molecule_featurizers/mat_featurizer.py:68]: Typo: "object"

File: deepchem/feat/molecule_featurizers/mat_featurizer.py [hunk 8]
```diff
@@ -84,25 +216,26 @@ def _featurize(self, datapoint: RDKitMol, **kwargs) -> np.ndarray:
 
     Returns
     -------
-    np.ndarray: A concatenated matrix consisting of node_features, adjacency_matrix and distance_matrix.
+    MATEncoding: A MATEncoding dataclass instance consisting of processed node_features, adjacency_matrix and distance_matrix.
```
Comment (rbharath) [inline deepchem/feat/molecule_featurizers/mat_featurizer.py:219]: Should be
```
MATEncoding
  A MATEncoding dataclass instance...
```

**Review by rbharath — COMMENTED** [review]
@atreyamaj I think this is basically ready to merge in! A couple of minor typos to address and we should be good to

**Comment by rbharath** [comment]
I'm going to go ahead and merge this and we can address the typos in #2676

**Pull request merged**